Fix cloud run deferrable job validation

BharathPESU opened the pull request:
Fixes a critical bug in [CloudRunExecuteJobOperator](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) where jobs with [deferrable=True](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) are incorrectly marked as successful when Cloud Run Jobs are canceled or have failed tasks.

Problem
When using [CloudRunExecuteJobOperator](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) with [deferrable=True](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), the operator would mark tasks as successful even when:

Jobs were canceled - Not all tasks completed execution
Tasks failed - Some tasks within the job failed
Malformed trigger events - Missing required fields could cause KeyError or produce unclear error messages
The root cause was that the trigger only checked [operation.done](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) without validating the actual execution state (task completion counts), and the operator didn't defensively validate the event payload.

Solution
This PR implements a three-layer defensive validation approach:

1. Trigger Layer ([cloud_run.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) trigger)
Extracts execution details ([task_count](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [succeeded_count](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [failed_count](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) from the operation response using protobuf [Unpack()](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Validates [Unpack()](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) return value and raises clear exceptions if deserialization fails
Includes operation name and job name in error messages for debugging
2. Operator Layer ([cloud_run.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) operator)
Defensive event validation: Uses [event.get()](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to prevent KeyError on malformed events
Execution state validation:
Checks [succeeded_count + failed_count == task_count](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to detect canceled jobs
Checks [failed_count > 0](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to detect failed tasks
Validates presence of required execution fields
Improved error messages: Replaces None/empty error details with meaningful defaults ("Unknown", "Unknown error")
3. Test Coverage
Added comprehensive test coverage for:

✅ Successful job execution with all tasks completed
✅ Canceled jobs (incomplete task execution)
✅ Jobs with failed tasks
✅ Malformed events missing status field
✅ Malformed events missing execution fields
✅ Operation failures with missing error details (None values)
✅ Protobuf Unpack() failure scenarios
Changes
Modified Files:

[cloud_run.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[cloud_run.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[test_cloud_run.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[test_cloud_run.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Statistics:

2 files changed in source code
2 test files enhanced
236 insertions total across both commits
4 deletions
Testing
All changes include comprehensive unit test coverage. Tests verify:

Success path with valid execution details
Failure detection for canceled jobs
Failure detection for jobs with failed tasks
Error handling for malformed trigger events
Meaningful error messages for missing operation error details
Protobuf Unpack() failure handling
Example Error Messages
Before this fix:

After this fix:

Checklist
 Bug fix (non-breaking change which fixes an issue)
 New tests added to cover the changes
 All tests pass locally
 Follows code style guidelines
 Includes appropriate error handling
 Error messages are clear and actionable
Related Issues
Note: This fix prevents false positives in production environments where canceled or partially completed Cloud Run Jobs would be incorrectly reported as successful, potentially leading to data inconsistencies or missed error conditions.

Feel free to modify this description based on any specific issue numbers or additional context you want to include!

BharathPESU committed:
Fix CloudRunExecuteJobOperator deferrable mode incorrectly marking canceled jobs as successful

When using CloudRunExecuteJobOperator with deferrable=True, canceled Cloud Run Jobs
were incorrectly marked as successful because the trigger only checked operation.done
without validating the actual execution state.

Changes:
- Trigger: Extract execution details (task_count, succeeded_count, failed_count) from
  operation.response using protobuf Unpack with proper error handling
- Operator: Add defensive validation for event dictionary and execution state to detect
  canceled jobs (incomplete execution) and failed tasks before marking success
- Tests: Add comprehensive test coverage for success, canceled jobs, failed tasks,
  malformed events, and Unpack failure scenarios

This ensures that only jobs where all tasks completed successfully are marked as
successful, preventing false positives for canceled or partially completed jobs.

BharathPESU committed:
Improve error handling for operation failures with missing error details

Previously, when operation_error_code or operation_error_message were missing or None,
the exception message would display 'None' literally, making error messages unclear.

Changes:
- Add validation to check if both error_code and error_message are missing/falsy,
  and raise a clear 'Operation failed with unknown error' message
- Replace None/empty error_code with 'Unknown' default
- Replace empty error_message with 'Unknown error' default
- Add comprehensive test coverage for edge cases:
  * Missing both error details
  * None error code with valid message
  * Empty error message with valid code

This ensures error messages are always clear and actionable for debugging.

VladaZakharova commented on providers/google/src/airflow/providers/google/cloud/triggers/cloud_run.py:
```diff
@@ -116,10 +116,25 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
                     )
                 else:
+                    # Parse the Execution object from operation.response to get execution details
+                    from google.cloud.run_v2.types import Execution
```
can you please move this import to the top of the file?

BharathPESU replied:
ok ,i will do it

VladaZakharova commented on providers/google/src/airflow/providers/google/cloud/triggers/cloud_run.py:
```diff
@@ -116,10 +116,25 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
+                            f"Failed to unpack Execution from operation response. "
+                            f"Operation: {self.operation_name}, Job: {self.job_name}"
+                        )
+                        self.log.error(error_msg)
```
I am not sure we need to output this error twice: in  logs and in raised exception. you can leave only one

BharathPESU committed:
Move Execution import to top-level and remove redundant error logging

- Move 'from google.cloud.run_v2.types import Execution' from inside the async
  function to the top-level imports for better code organization
- Remove redundant self.log.error() call before raising AirflowException to avoid
  duplicate error output (feedback from repo maintainer)
- The exception message already provides sufficient context for debugging

BharathPESU reviewed (commented):
if any changes needed ,tell me
